nvtabular/framework_utils/tensorflow/layers/embedding.py

```diff
@@ -0,0 +1,304 @@
+import tensorflow as tf
+from tensorflow.python.feature_column import feature_column_v2 as fc
+
+
+def _sort_columns(feature_columns):
+    return sorted(feature_columns, key=lambda col: col.name)
+
+
+def _validate_numeric_column(feature_column):
+    if len(feature_column.shape) > 1:
+        return "Matrix numeric features are not allowed, " "found feature {} with shape {}".format(
+            feature_column.key, feature_column.shape
+        )
+    elif feature_column.shape[0] != 1:
+        return "Vector numeric features are not allowed, " "found feature {} with shape {}".format(
+            feature_column.key, feature_column.shape
+        )
+
+
+def _validate_categorical_column(feature_column):
+    if not isinstance(feature_column, fc.IdentityCategoricalColumn):
+        return (
+            "Only acceptable categorical columns for feeding "
+            "embeddings are identity, found column {} of type {}. "
+            "Consider using NVTabular online preprocessing to perform "
+            "categorical transformations".format(feature_column.name, type(feature_column).__name__)
+        )
+
+
+def _validate_dense_feature_columns(feature_columns):
+    _errors = []
+    for feature_column in feature_columns:
+        if isinstance(feature_column, fc.CategoricalColumn):
+            if not isinstance(feature_column, fc.BucketizedColumn):
+                _errors.append(
+                    "All feature columns must be dense, found categorical "
+                    "column {} of type {}. Please wrap categorical columns "
+                    "in embedding or indicator columns before passing".format(
+                        feature_column.name, type(feature_column).__name__
+                    )
+                )
+            else:
+                _errors.append(
+                    "Found bucketized column {}. ScalarDenseFeatures layer "
+                    "cannot apply bucketization preprocessing. Consider using "
+                    "NVTabular to do preprocessing offline".format(feature_column.name)
+                )
+        elif isinstance(feature_column, (fc.EmbeddingColumn, fc.IndicatorColumn)):
+            _errors.append(_validate_categorical_column(feature_column.categorical_column))
+
+        elif isinstance(feature_column, fc.NumericColumn):
+            _errors.append(_validate_numeric_column(feature_column))
+
+    _errors = list(filter(lambda e: e is not None, _errors))
+    if len(_errors) > 0:
+        msg = "Found issues with columns passed to ScalarDenseFeatures:"
+        msg += "\n\t".join(_errors)
+        raise ValueError(_errors)
+
+
+def _validate_stack_dimensions(feature_columns):
+    dims = []
+    for feature_column in feature_columns:
+        try:
+            dimension = feature_column.dimension
+        except AttributeError:
+            dimension = feature_column.shape[0]
+        dims.append(dimension)
+
+    dim0 = dims[0]
+    if not all([dim == dim0 for dim in dims[1:]]):
```

Comment:
nitpick: you don't need to create a list here - using a generator will avoid allocating new memory:
```suggestion
    if not all(dim == dim0 for dim in dims[1:]):
```

Reply:
good call, fixed with e4e6f0c018dfd1c69ba4026c0f117f1851941b44